homeassistant/components/mqtt/models.py

```diff
@@ -369,17 +375,54 @@ def __init__(self) -> None:
         """Register topic."""
         self.subscribe_calls: dict[str, Entity] = {}
 
+    def _check_state_too_long(
+        self, entity: Entity, entity_id: str, msg: MQTTMessage
+    ) -> bool:
+        """Check if the processed state is too long."""
+        try:
+            state: str = entity.state  # type: ignore[assignment]
+            state_length = len(state)
+        except (TypeError, ValueError):
+            # When the state is not a string or is invalid, we
+            # ignore this at this point, there errors will be handled
+            # when we will try to write the state.
+            return False
+        if state_length > MAX_LENGTH_STATE_STATE:
+            _LOGGER.error(
+                "Cannot update state for entity %s after processing "
+                "payload on topic %s. The requested state (%s) exceeds "
+                "the maximum allowed length (%s). Fall back to "
+                "%s, failed state: %s",
+                entity_id,
+                msg.topic,
+                state_length,
+                MAX_LENGTH_STATE_STATE,
+                STATE_UNKNOWN,
+                state[:8192],
+            )
+            entity.hass.states.async_set(
+                entity_id,
+                STATE_UNKNOWN,
+                {},
+                entity.force_update,
+            )
+            return True
+
+        return False
+
     @callback
     def process_write_state_requests(self, msg: MQTTMessage) -> None:
         """Process the write state requests."""
         while self.subscribe_calls:
-            _, entity = self.subscribe_calls.popitem()
+            entity_id, entity = self.subscribe_calls.popitem()
+            if self._check_state_too_long(entity, entity_id, msg):
+                return
```

Comment:
I think its better to catch it closer to where the data comes in and set it to unknown there instead

Reply:
Moved to `sensor` and `text` platform code.